File: 129. Sum Root to Leaf Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // DFS with string, O(Node), 9.12%, ok
    int sumNumbers(TreeNode* root) {
        if(!root) return 0;
        vector<string> ans;
        dfs_string(root, ans, "");
        int ret = 0;
        for(int i = 0; i < ans.size(); i++) ret += stoi(ans[i]);
        return ret;
    }

    void dfs_string(TreeNode* root, vector<string>& ans, string sum){
        if(!root) return;
        if(!root->left && !root->right) ans.push_back(sum + to_string(root->val));
        if(root->left) dfs_string(root->left, ans, sum + to_string(root->val));
        if(root->right) dfs_string(root->right, ans, sum + to_string(root->val));
        return;
    }


    // DFS with calculation, O(Node), 9.12%, ok
    int sumNumbers(TreeNode* root) {
        if(!root) return 0;
        vector<long> ans;
        dfs_num(root, ans, 0);
        int ret = 0;
        for(int i = 0; i < ans.size(); i++) ret += ans[i];
        return ret;
    }

    void dfs_num(TreeNode* root, vector<long>& ans, long sum){
        if(!root) return;
        if(!root->left && !root->right) ans.push_back(sum * 10 + root->val);
        if(root->left) dfs_num(root->left, ans, sum * 10 + root->val);
        if(root->right) dfs_num(root->right, ans, sum * 10 + root->val);
        return;
    }
};

File: 129. Sum Root to Leaf Numbers.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
    # revursive dfs: O(N) time & space
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.ret = 0
        self.dfs(root, 0)
        return self.ret
       
    def dfs(self, node, currSum):
        if not node: return
        if not node.left and not node.right: self.ret += currSum*10 + node.val
        self.dfs(node.left, currSum*10 + node.val)
        self.dfs(node.right, currSum*10 + node.val)

    # dfs + stack: O(N) time & space
    def sumNumbers(self, root):
        if not root: return 0
        stack, ret = [(root, root.val)], 0
        while stack:
            node, val = stack.pop()
            if not node.left and not node.right: ret += val
            if node.right: stack.append((node.right, val*10 + node.right.val))
            if node.left: stack.append((node.left, val*10 + node.left.val))
        return ret
